File: backend/app/core/sentry.py
"""
Configuration Sentry pour le tracking des erreurs
"""

import os
from typing import Optional, Dict, Any

from app.core.logging import get_logger

logger = get_logger(__name__)

# Flag pour vérifier si Sentry est disponible
SENTRY_AVAILABLE = False

try:
    import sentry_sdk
    from sentry_sdk.integrations.fastapi import FastApiIntegration
    from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
    from sentry_sdk.integrations.redis import RedisIntegration
    from sentry_sdk.integrations.logging import LoggingIntegration
    SENTRY_AVAILABLE = True
except ImportError:
    logger.warning("sentry-sdk not installed. Error tracking disabled.")


def init_sentry(
    dsn: Optional[str] = None,
    environment: str = "development",
    release: Optional[str] = None,
    traces_sample_rate: float = 0.1,
    profiles_sample_rate: float = 0.1,
) -> bool:
    """
    Initialise Sentry pour le tracking des erreurs

    Args:
        dsn: Sentry DSN (Data Source Name). Si None, utilise SENTRY_DSN env var
        environment: Environnement (development, staging, production)
        release: Version de l'application
        traces_sample_rate: Taux d'échantillonnage des traces (0.0 à 1.0)
        profiles_sample_rate: Taux d'échantillonnage des profils (0.0 à 1.0)

    Returns:
        True si Sentry a été initialisé avec succès, False sinon
    """
    if not SENTRY_AVAILABLE:
        logger.info("Sentry SDK not available, skipping initialization")
        return False

    # Récupérer le DSN depuis les variables d'environnement si non fourni
    sentry_dsn = dsn or os.environ.get("SENTRY_DSN")

    if not sentry_dsn:
        logger.info("No Sentry DSN configured, error tracking disabled")
        return False

    try:
        sentry_sdk.init(
            dsn=sentry_dsn,
            environment=environment,
            release=release,
            traces_sample_rate=traces_sample_rate,
            profiles_sample_rate=profiles_sample_rate,

            # Intégrations
            integrations=[
                FastApiIntegration(
                    transaction_style="endpoint"
                ),
                SqlalchemyIntegration(),
                RedisIntegration(),
                LoggingIntegration(
                    level=None,  # Ne pas capturer les logs comme breadcrumbs
                    event_level=None  # Ne pas créer d'events pour les logs
                ),
            ],

            # Configuration additionnelle
            send_default_pii=False,  # Ne pas envoyer de données personnelles
            attach_stacktrace=True,
            max_breadcrumbs=50,

            # Filtrer les erreurs non importantes
            before_send=filter_events,

            # Filtrer les transactions (performance)
            before_send_transaction=filter_transactions,
        )

        logger.info(f"Sentry initialized for environment: {environment}")
        return True

    except Exception as e:
        logger.error(f"Failed to initialize Sentry: {e}")
        return False


def filter_events(event: Dict[str, Any], hint: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Filtre les événements avant envoi à Sentry

    Permet d'exclure certaines erreurs ou de modifier les événements
    """
    # Récupérer l'exception si disponible
    if "exc_info" in hint:
        exc_type, exc_value, tb = hint["exc_info"]

        # Ignorer certaines erreurs communes non critiques
        ignored_exceptions = [
            "ConnectionResetError",
            "BrokenPipeError",
            "asyncio.CancelledError",
        ]

        if exc_type.__name__ in ignored_exceptions:
            return None

        # Ignorer les erreurs 404
        if hasattr(exc_value, "status_code") and exc_value.status_code == 404:
            return None

    # Supprimer les données sensibles
    if "request" in event:
        request = event["request"]

        # Supprimer les headers sensibles
        if "headers" in request:
            sensitive_headers = ["authorization", "cookie", "x-api-key"]
            request["headers"] = {
                k: "[FILTERED]" if k.lower() in sensitive_headers else v
                for k, v in request["headers"].items()
            }

        # Supprimer les données de formulaire sensibles
        if "data" in request and isinstance(request["data"], dict):
            sensitive_fields = ["password", "mot_de_passe", "token", "secret"]
            request["data"] = {
                k: "[FILTERED]" if any(s in k.lower() for s in sensitive_fields) else v
                for k, v in request["data"].items()
            }

    return event


def filter_transactions(event: Dict[str, Any], hint: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Filtre les transactions de performance avant envoi à Sentry

    Permet d'exclure certains endpoints du monitoring de performance
    """
    # Ignorer les endpoints de health check et metrics
    transaction_name = event.get("transaction", "")
    ignored_endpoints = [
        "/health",
        "/health/live",
        "/health/ready",
        "/metrics",
        "/docs",
        "/redoc",
        "/openapi.json",
    ]

    for endpoint in ignored_endpoints:
        if transaction_name.endswith(endpoint):
            return None

    return event


def capture_exception(exception: Exception, **kwargs) -> Optional[str]:
    """
    Capture une exception et l'envoie à Sentry

    Args:
        exception: L'exception à capturer
        **kwargs: Données additionnelles (tags, extra, etc.)

    Returns:
        L'ID de l'événement Sentry ou None si non disponible
    """
    if not SENTRY_AVAILABLE:
        return None

    try:
        with sentry_sdk.push_scope() as scope:
            # Ajouter les tags
            if "tags" in kwargs:
                for key, value in kwargs["tags"].items():
                    scope.set_tag(key, value)

            # Ajouter les données extra
            if "extra" in kwargs:
                for key, value in kwargs["extra"].items():
                    scope.set_extra(key, value)

            # Ajouter le contexte utilisateur
            if "user" in kwargs:
                scope.set_user(kwargs["user"])

            return sentry_sdk.capture_exception(exception)
    except Exception as e:
        logger.error(f"Failed to capture exception in Sentry: {e}")
        return None


def capture_message(message: str, level: str = "info", **kwargs) -> Optional[str]:
    """
    Capture un message et l'envoie à Sentry

    Args:
        message: Le message à capturer
        level: Niveau du message (debug, info, warning, error, fatal)
        **kwargs: Données additionnelles

    Returns:
        L'ID de l'événement Sentry ou None si non disponible
    """
    if not SENTRY_AVAILABLE:
        return None

    try:
        with sentry_sdk.push_scope() as scope:
            if "tags" in kwargs:
                for key, value in kwargs["tags"].items():
                    scope.set_tag(key, value)

            if "extra" in kwargs:
                for key, value in kwargs["extra"].items():
                    scope.set_extra(key, value)

            return sentry_sdk.capture_message(message, level=level)
    except Exception as e:
        logger.error(f"Failed to capture message in Sentry: {e}")
        return None


def set_user_context(user_id: str, email: Optional[str] = None, role: Optional[str] = None):
    """
    Définit le contexte utilisateur pour les événements Sentry

    Args:
        user_id: ID de l'utilisateur
        email: Email de l'utilisateur (optionnel)
        role: Rôle de l'utilisateur (optionnel)
    """
    if not SENTRY_AVAILABLE:
        return

    try:
        sentry_sdk.set_user({
            "id": user_id,
            "email": email,
            "role": role,
        })
    except Exception as e:
        logger.error(f"Failed to set user context in Sentry: {e}")


def clear_user_context():
    """Efface le contexte utilisateur"""
    if not SENTRY_AVAILABLE:
        return

    try:
        sentry_sdk.set_user(None)
    except Exception as e:
        logger.error(f"Failed to clear user context in Sentry: {e}")

File: frontend/src/utils/sentry.ts
/**
 * Configuration Sentry pour le tracking des erreurs frontend
 *
 * Installation:
 *   npm install @sentry/react @sentry/tracing
 *
 * Variables d'environnement:
 *   VITE_SENTRY_DSN - DSN Sentry pour le projet
 *   VITE_ENVIRONMENT - Environnement (development, staging, production)
 */

// Types pour Sentry (en attendant l'installation du package)
interface SentryConfig {
  dsn: string;
  environment: string;
  release?: string;
  tracesSampleRate: number;
  replaysSessionSampleRate: number;
  replaysOnErrorSampleRate: number;
}

interface SentryUser {
  id: string;
  email?: string;
  username?: string;
}

// Vérifier si Sentry est disponible
let sentryAvailable = false;
let Sentry: typeof import('@sentry/react') | null = null;

/**
 * Initialise Sentry pour le frontend
 */
export async function initSentry(): Promise<boolean> {
  const dsn = import.meta.env.VITE_SENTRY_DSN;

  if (!dsn) {
    console.info('Sentry DSN not configured, error tracking disabled');
    return false;
  }

  try {
    // Import dynamique de Sentry
    Sentry = await import('@sentry/react');

    const config: SentryConfig = {
      dsn,
      environment: import.meta.env.VITE_ENVIRONMENT || 'development',
      release: import.meta.env.VITE_APP_VERSION,
      tracesSampleRate: import.meta.env.PROD ? 0.1 : 1.0,
      replaysSessionSampleRate: 0.1,
      replaysOnErrorSampleRate: 1.0,
    };

    Sentry.init({
      dsn: config.dsn,
      environment: config.environment,
      release: config.release,
      tracesSampleRate: config.tracesSampleRate,

      // Intégrations
      integrations: [
        Sentry.browserTracingIntegration(),
        Sentry.replayIntegration({
          maskAllText: true,
          blockAllMedia: true,
        }),
      ],

      // Ne pas envoyer de données personnelles
      sendDefaultPii: false,

      // Filtrer les erreurs
      beforeSend(event, hint) {
        return filterEvent(event, hint);
      },

      // Filtrer les breadcrumbs
      beforeBreadcrumb(breadcrumb) {
        return filterBreadcrumb(breadcrumb);
      },
    });

    sentryAvailable = true;
    console.info(`Sentry initialized for environment: ${config.environment}`);
    return true;
  } catch (error) {
    console.error('Failed to initialize Sentry:', error);
    return false;
  }
}

/**
 * Filtre les événements avant envoi
 */
function filterEvent(event: Record<string, unknown>, hint: Record<string, unknown>): Record<string, unknown> | null {
  const error = hint.originalException as Error | undefined;

  // Ignorer certaines erreurs communes
  if (error) {
    const ignoredMessages = [
      'ResizeObserver loop',
      'Network request failed',
      'Load failed',
      'ChunkLoadError',
    ];

    const message = error.message || '';
    if (ignoredMessages.some((ignored) => message.includes(ignored))) {
      return null;
    }
  }

  // Supprimer les données sensibles des URLs
  if (event.request && typeof event.request === 'object') {
    const request = event.request as Record<string, unknown>;
    if (typeof request.url === 'string') {
      request.url = sanitizeUrl(request.url);
    }
  }

  return event;
}

/**
 * Filtre les breadcrumbs
 */
function filterBreadcrumb(breadcrumb: Record<string, unknown>): Record<string, unknown> | null {
  // Ignorer les requêtes vers les endpoints de health check
  if (breadcrumb.category === 'fetch' || breadcrumb.category === 'xhr') {
    const url = breadcrumb.data?.url as string | undefined;
    if (url && (url.includes('/health') || url.includes('/metrics'))) {
      return null;
    }
  }

  return breadcrumb;
}

/**
 * Supprime les tokens et données sensibles des URLs
 */
function sanitizeUrl(url: string): string {
  try {
    const urlObj = new URL(url);
    const sensitiveParams = ['token', 'access_token', 'refresh_token', 'api_key', 'password'];

    sensitiveParams.forEach((param) => {
      if (urlObj.searchParams.has(param)) {
        urlObj.searchParams.set(param, '[FILTERED]');
      }
    });

    return urlObj.toString();
  } catch {
    return url;
  }
}

/**
 * Capture une exception
 */
export function captureException(
  error: Error,
  context?: {
    tags?: Record<string, string>;
    extra?: Record<string, unknown>;
    user?: SentryUser;
  }
): string | undefined {
  if (!sentryAvailable || !Sentry) {
    console.error('Sentry not available, logging error locally:', error);
    return undefined;
  }

  return Sentry.withScope((scope) => {
    if (context?.tags) {
      Object.entries(context.tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (context?.extra) {
      Object.entries(context.extra).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
    }

    if (context?.user) {
      scope.setUser(context.user);
    }

    return Sentry!.captureException(error);
  });
}

/**
 * Capture un message
 */
export function captureMessage(
  message: string,
  level: 'debug' | 'info' | 'warning' | 'error' | 'fatal' = 'info',
  context?: Record<string, unknown>
): string | undefined {
  if (!sentryAvailable || !Sentry) {
    console.log(`[${level}] ${message}`, context);
    return undefined;
  }

  return Sentry.withScope((scope) => {
    if (context) {
      Object.entries(context).forEach(([key, value]) => {
        scope.setExtra(key, value);
      });
    }
    return Sentry!.captureMessage(message, level);
  });
}

/**
 * Définit le contexte utilisateur
 */
export function setUser(user: SentryUser | null): void {
  if (!sentryAvailable || !Sentry) return;

  if (user) {
    Sentry.setUser(user);
  } else {
    Sentry.setUser(null);
  }
}

/**
 * Ajoute un breadcrumb personnalisé
 */
export function addBreadcrumb(
  message: string,
  category: string,
  level: 'debug' | 'info' | 'warning' | 'error' = 'info',
  data?: Record<string, unknown>
): void {
  if (!sentryAvailable || !Sentry) return;

  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  });
}

/**
 * Wrapper ErrorBoundary Sentry
 * Utiliser à la place de notre ErrorBoundary personnalisé si Sentry est disponible
 */
export function getSentryErrorBoundary(): typeof import('@sentry/react').ErrorBoundary | null {
  if (!sentryAvailable || !Sentry) return null;
  return Sentry.ErrorBoundary;
}
